File: Solution with Bitset for recording ancestors.java

import java.util.ArrayList;
import java.util.BitSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Solution {
    
    private int numberOfNodes;
    
    public List<List<Integer>> getAncestors(int numberOfNodes, int[][] edges) {
        this.numberOfNodes = numberOfNodes;
        int[] incomingEdgesPerNode = createArrayIncomingEdgesPerNode(edges);
        List<Integer>[] directedAcyclicGraph = createDirectedAcyclicGraph(edges);
        
        return findAncestorsPerNodeWithTopologicalSort(directedAcyclicGraph, incomingEdgesPerNode);
    }
    
    private List<List<Integer>> findAncestorsPerNodeWithTopologicalSort(List<Integer>[] directedAcyclicGraph, int[] incomingEdgesPerNode) {
        Queue<Integer> queue = createQueueNodesWithoutIncomingEdges(incomingEdgesPerNode);
        BitSet[] ancestors = new BitSet[numberOfNodes];
        for (int nodeID = 0; nodeID < numberOfNodes; ++nodeID) {
            ancestors[nodeID] = new BitSet(numberOfNodes);
        }
        
        while (!queue.isEmpty()) {
            int current = queue.poll();
            
            for (int next : directedAcyclicGraph[current]) {
                updateAncestors(current, next, ancestors);
                if (--incomingEdgesPerNode[next] == 0) {
                    queue.add(next);
                }
            }
        }
        return createListAllAncestorsPerNode(ancestors);
    }
    
    private int[] createArrayIncomingEdgesPerNode(int[][] edges) {
        int[] incomingEdgesPerNode = new int[numberOfNodes];
        for (int[] edge : edges) {
            int to = edge[1];
            ++incomingEdgesPerNode[to];
        }
        return incomingEdgesPerNode;
    }
    
    private List<Integer>[] createDirectedAcyclicGraph(int[][] edges) {
        List<Integer>[] directedAcyclicGraph = new ArrayList[numberOfNodes];
        for (int nodeID = 0; nodeID < numberOfNodes; ++nodeID) {
            directedAcyclicGraph[nodeID] = new ArrayList<>();
        }
        
        for (int[] edge : edges) {
            int from = edge[0];
            int to = edge[1];
            directedAcyclicGraph[from].add(to);
        }
        return directedAcyclicGraph;
    }
    
    private Queue<Integer> createQueueNodesWithoutIncomingEdges(int[] incomingEdgesPerNode) {
        Queue<Integer> queue = new LinkedList<>();
        for (int nodeID = 0; nodeID < numberOfNodes; ++nodeID) {
            if (incomingEdgesPerNode[nodeID] == 0) {
                queue.add(nodeID);
            }
        }
        return queue;
    }
    
    private List<List<Integer>> createListAllAncestorsPerNode(BitSet[] ancestors) {
        List<List<Integer>> allAncestorsPerNode = new ArrayList<>();
        
        for (int nodeID = 0; nodeID < numberOfNodes; ++nodeID) {
            List<Integer> ancestorsForCurrentNode = new ArrayList<>();
            
            for (int candidateAncestorID = 0; candidateAncestorID < numberOfNodes; ++candidateAncestorID) {
                if (ancestors[nodeID].get(candidateAncestorID)) {
                    ancestorsForCurrentNode.add(candidateAncestorID);
                }
            }
            allAncestorsPerNode.add(ancestorsForCurrentNode);
        }
        return allAncestorsPerNode;
    }
    
    private void updateAncestors(int current, int next, BitSet[] ancestors) {
        ancestors[next].or(ancestors[current]);
        ancestors[next].set(current);
    }
}

File: Solution with Bitset for recording ancestors.kt

import java.util.*
import kotlin.collections.ArrayList

class Solution {

    private var numberOfNodes: Int = 0

    fun getAncestors(numberOfNodes: Int, edges: Array<IntArray>): List<List<Int>> {
        this.numberOfNodes = numberOfNodes
        val incomingEdgesPerNode = createArrayIncomingEdgesPerNode(edges)
        val directedAcyclicGraph = createDirectedAcyclicGraph(edges)

        return findAncestorsPerNodeWithTopologicalSort(directedAcyclicGraph, incomingEdgesPerNode)
    }

    private fun findAncestorsPerNodeWithTopologicalSort(directedAcyclicGraph: ArrayList<ArrayList<Int>>, incomingEdgesPerNode: IntArray): List<List<Int>> {
        val queue = createQueueNodesWithoutIncomingEdges(incomingEdgesPerNode)
        val ancestors = ArrayList<BitSet>(numberOfNodes)
        for (nodeID in 0..<numberOfNodes) {
            ancestors.add(BitSet(numberOfNodes))
        }

        while (!queue.isEmpty()) {
            val current = queue.poll()

            for (next in directedAcyclicGraph[current]) {
                updateAncestors(current, next, ancestors)
                if (--incomingEdgesPerNode[next] == 0) {
                    queue.add(next)
                }
            }
        }
        return createListAllAncestorsPerNode(ancestors)
    }

    private fun createArrayIncomingEdgesPerNode(edges: Array<IntArray>): IntArray {
        val incomingEdgesPerNode = IntArray(numberOfNodes)
        for (edge in edges) {
            val to = edge[1]
            ++incomingEdgesPerNode[to]
        }
        return incomingEdgesPerNode
    }

    private fun createDirectedAcyclicGraph(edges: Array<IntArray>): ArrayList<ArrayList<Int>> {
        val directedAcyclicGraph = ArrayList<ArrayList<Int>>(numberOfNodes)
        for (nodeID in 0..<numberOfNodes) {
            directedAcyclicGraph.add(ArrayList<Int>())
        }

        for ((from, to) in edges) {
            directedAcyclicGraph[from].add(to)
        }
        return directedAcyclicGraph
    }

    private fun createQueueNodesWithoutIncomingEdges(incomingEdgesPerNode: IntArray): Queue<Int> {
        val queue = LinkedList<Int>()
        for (nodeID in 0..<numberOfNodes) {
            if (incomingEdgesPerNode[nodeID] == 0) {
                queue.add(nodeID)
            }
        }
        return queue
    }

    private fun createListAllAncestorsPerNode(ancestors: ArrayList<BitSet>): List<List<Int>> {
        val allAncestorsPerNode = ArrayList<List<Int>>()

        for (nodeID in 0..<numberOfNodes) {
            val ancestorsForCurrentNode = ArrayList<Int>()

            for (candidateAncestorID in 0..<numberOfNodes) {
                if (ancestors[nodeID][candidateAncestorID]) {
                    ancestorsForCurrentNode.add(candidateAncestorID)
                }
            }
            allAncestorsPerNode.add(ancestorsForCurrentNode)
        }
        return allAncestorsPerNode
    }

    private fun updateAncestors(current: Int, next: Int, ancestors: ArrayList<BitSet>) {
        ancestors[next].or(ancestors[current])
        ancestors[next].set(current)
    }
}
